### ISSUE TITLE
sql: preparing statements takes leases and relies on running a statement to release them

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#7648]
(Found by @dt and myself)

Steps to repro:
1. start new virgin db
2. run the attached [test SQL code](https://github.com/cockroachdb/cockroach/files/350211/test.txt) once: `cat test.txt | ./cockroach sql`
3. stop the db (Ctrl+C twice, not `cockroach quit`); then restart the db
4. repeat step #2.

This will cause the db to fail with a loop on `waitForOneVersion`:

```
I160706 15:22:41.873651 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:42.083795 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:42.296096 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:42.490031 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:42.697195 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:42.906908 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:43.109133 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:43.339883 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:43.562171 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:43.750974 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:43.925267 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:44.147329 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
I160706 15:22:44.348866 sql/lease.go:249  publish (count leases): descID=51 name=t1 version=16 count=1
```

**Comment by vivekmenezes** [comment]
@dt founds another way to repro
```
[15:35:19] ~/code/go/src/github.com/cockroachdb/cockroach $ g diff
diff --git a/sql/pgwire_test.go b/sql/pgwire_test.go
index dbf3a64..f21588e 100644
--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -780,6 +780,7 @@ func TestPGPreparedExec(t *testing.T) {
                    []byte{},
                    time.Time{}, // date
                    time.Time{}, // timestamp
+                   nil,
                    time.Hour.String(),
                    true,
                    "0.0", // decimal
```

**Comment by andreimatei** [comment]
The problem here is that a lease taken while **preparing** a statement is only released after subsequently running a statement (or when the txn commits).
This results in leases being held by the planner after preparing a statement, even though there's no reason for that. Even worse, because of #7658, if the client goes away and never runs any commands, the lease is not released.

**Comment by andreimatei** [comment]
I think we can do one of two things here, although I don't think I'll do either for the time being (but I'll fix #7658).

1) we can release the leases that preparing takes after the prepare is done. The problem is that leases are simply accumulated on the planner without knowledge of what operation acquired them. We could in principle do the preparing using a fresh planner, but there's a cyclical dependency between a planner and a session (although there's a TODO in the planner to eliminate it). We have to be careful with transactions, though. This temporary planner must acquire leases in the context of the ongoing txn, is any.
2) we can skip acquiring leases at all for preparing statements; they aren't actually needed for validating a query - which is (the only thing) that preparing a statement currently does. Maybe we can provide another implementation for @knz's `SchemaAccessor` interface and prove that interface to be technically useful.

I'll punt on all of these, hoping that preparing statements is enhanced and better integrated with leases - seems likely that prepared statements will cache an execution plan at some point, so they'll need to be tied to schema versions.

**Comment by spencerkimball** [comment]
@andreimatei I'm intuiting this will not be addressed for 1.0. If you believe it should be, please reinstate the 1.0 milestone.

**Comment by vivekmenezes** [comment]
I'm closing this issue because PREPARE no longer holds on to a table descriptor lease.
